Using-the-Compiler-API.md

```diff
@@ -433,4 +433,154 @@ function compile(sourceFiles: string[], moduleSearchLocations: string[]): void {
 }
-```
\ No newline at end of file
+```
+
+### Using the Type Checker
+
+In this example we will walk the AST and use the checker to serialize class information.
+We'll use the type checker to get symbol and type information, while grabbing JSDoc comments for classes, their constructors, and respective constructor parameters.
+
+```ts
+/// <reference path="typings/node/node.d.ts" />
+
+import * as ts from "typescript";
+import * as fs from "fs";
+
+interface DocEntry {
+    name?: string,
+    fileName?: string,
+    documentation?: string,
+    type?: string,
+    constructors?: DocEntry[],
+    parameters?: DocEntry[],
+    returnType?: string
+};
+
+/** Generate documention for all classes in a set of .ts files */
+function generateDocumentation(fileNames: string[], options: ts.CompilerOptions): void {
+    // Build a program using the set of root file names in fileNames
+    let program = ts.createProgram(fileNames, options);
+    
+    // Get the checker, we will use it to find more about classes
+    let checker = program.getTypeChecker();
+
+    let output: DocEntry[] = [];
+
+    // Visit every sourceFile in the program    
+    for (const sourceFile of program.getSourceFiles()) {
+        // Walk the tree to search for classes
+        ts.forEachChild(sourceFile, visit);
+    });
+
+    // print out the doc
+    fs.writeFileSync("classes.json", JSON.stringify(output, undefined, 4));
+
+    /** Serialize a symbol into a json object */    
+    function serializeSymbol(symbol: ts.Symbol): DocEntry {
+        return {
+            name: symbol.getName(),
+            documentation: ts.displayPartsToString(symbol.getDocumentationComment()),
+            type: checker.typeToString(checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration))
+        };
+    }
+
+    /** Serialize a class symbol infomration */
+    function serializeClass(symbol: ts.Symbol) {
+        let details = serializeSymbol(symbol);
+        
+        // Get the construct signatures            
+        let constructorType = checker.getTypeOfSymbolAtLocation(symbol, symbol.valueDeclaration);
+        details.constructors = constructorType.getConstructSignatures().map(serializeSignature);
+        return details;
+    }
+
+    /** Serialize a signature (call or construct) */    
+    function serializeSignature(signature: ts.Signature) {
+        return {
+            paramters: signature.parameters.map(serializeSymbol),
+            returnType: checker.typeToString(signature.getReturnType()),
+            documentation: ts.displayPartsToString(signature.getDocumentationComment())
+        };
+    }
+
+    /** True if this is visible outside this file, false otherwise */
+    function isNodeExported(node: ts.Node): boolean {
+        return (node.flags & ts.NodeFlags.Export) !== 0 || (node.parent && node.parent.kind === ts.SyntaxKind.SourceFile);
+    }
+
+    /** vist nodes finding exported classes */    
+    function visit(node: ts.Node) {
```

Comment:
This is so far down compared to the actual call to `visit`. Order these such that they reflect when the calls occur, from earlier on top, to later on the bottom.